Support GRU module quantization in Pytorch

lly-zero-one opened the pull request:
Differential Revision: D16037021

jamesr66a commented on test/test_jit.py:
```diff
@@ -6295,6 +6295,77 @@ def forward(self, x, hiddens):
+        # export_import_wrapper = self.getExportImportCopyWithPacking(wrapper)
+        # ei_out, ei_hid = export_import_wrapper(x)
+        # torch.testing.assert_allclose(ei_out, output)
+        # torch.testing.assert_allclose(ei_hid, final_hiddens)
```
Why are these commented?

llyfacebook replied:
Need to wait for the fix of script bug ("Optional type does not work properly in script_method during tracing").

jamesr66a commented on test/test_jit.py:
```diff
@@ -6295,6 +6295,77 @@ def forward(self, x, hiddens):
+                     'Quantized RNN requires FBGEMM. FBGEMM does not play'
+                     ' well with UBSAN at the moment, so we skip the test if'
+                     ' we are in a UBSAN environment.')
+    def test_gru_quantized(self):
```
Instead of making an entirely new test, can you just combine the LSTM and GRU tests like what was done for the cells:

https://github.com/pytorch/pytorch/blob/046c4589df15fc4e7b832d594d3d8e3a54cd47c9/test/test_jit.py#L6128

jamesr66a commented on torch/jit/quantized.py:
```diff
@@ -224,16 +227,16 @@ def __init__(self, other):
         self.bias = other.bias
         self.batch_first = other.batch_first
-        assert not self.batch_first
+        # assert not self.batch_first
```
Why?

jamesr66a reviewed (commented):
Looks okay, small comments

BIT-silence commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -747,7 +747,48 @@ std::tuple<Tensor, Tensor> NAME(                                               \
+    return std::make_tuple(output, hy);                                        \
+  }                                                                            \
+  check_device(_input, _params, hx);                                        \
+  auto input = batch_first ? _input.transpose(0, 1) : _input;                  \
```
If batch_first is false, there will be a copy for the input, right?

BIT-silence commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -747,7 +747,48 @@ std::tuple<Tensor, Tensor> NAME(                                               \
+  auto results = _rnn_impl_with_concat<CELL, FullLayer, FullBidirectionalLayer>( \
+          input, params, hx.unbind(0), num_layers, dropout_p, train, bidirectional); \
+  if (batch_first) {                                                           \
+    std::get<0>(results) = std::get<0>(results).transpose(0, 1);               \
```
```suggestion
    std::get<0>(results).transpose_(0, 1);              \
```

llyfacebook replied:
yes, make sense. A reference return.

llyfacebook replied:
seems tons of places need to be fixed!!

BIT-silence commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -747,7 +747,48 @@ std::tuple<Tensor, Tensor> NAME(                                               \
+            _params, has_biases, num_layers, dropout_p, train, bidirectional); \
+    return std::make_tuple(output, hy);                                        \
+  }                                                                            \
+  PackedSequence input { data, batch_sizes };                                  \
```
```suggestion
  PackedSequence input = {data, batch_sizes};                                  \
```
https://abseil.io/tips/88

BIT-silence commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -747,7 +747,48 @@ std::tuple<Tensor, Tensor> NAME(                                               \
 
+#define ONE_HIDDEN_QRNN(NAME, CELL)                                             \
+std::tuple<Tensor, Tensor> NAME(                                               \
+      const Tensor& _input, const Tensor& hx,                                  \
```
Consider make it a better format?

MaigoAkisame commented on torch/jit/quantized.py:
```diff
@@ -485,4 +561,6 @@ def quantize_rnn_modules(module):
     if isinstance(mod, torch.nn.LSTM):
         return QuantizedLSTM(mod)
+    if isinstance(mod, torch.nn.GRU):
+        return QuantizedGRU(mod)
```
Line 562~565: The variable `mod` is not defined in this scope. Did you mean `module`?
(the same applies to Lines 546~547)

MaigoAkisame replied:
Line 562-565: The variable `mod` is not defined in this scope. Did you mean `module`?
(the same applies to Lines 546-547)

lly-zero-one committed:
Support GRU module quantization in Pytorch

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/22498

Reviewed By: BIT-silence

Differential Revision: D16037021

fbshipit-source-id: 6e30b0a76e7afebaf7b4de927df6ebaab735ca6b
